### PR TITLE
External locators feature [14627]

### PR DESCRIPTION
**Pull request by MiguelCompany**

### CODE REVIEW

File: docs/fastdds/xml_configuration/common.rst [hunk 1]
```diff
@@ -82,6 +83,43 @@ The following example shows the implementation of one locator of each transport
...
++---------------------+--------------------------------------------------------+--------------------+------------------+
+| ``cost``            | Communication cost relative to other locators on the   | ``uint8_t``        | 0                |
+|                     | same externality level.                                |                    |                  |
+|                     | Valid values: from 0 to 255.                           |                    |                  |
++---------------------+--------------------------------------------------------+--------------------+------------------+
+| ``mask``            | Number of significant bits on the LAN represented by   | ``uint8_t``        | 24               |
+|                     | the external locator.                                  |                    |                  |
+|                     | Valid values: from 1 to 31 (UDPv4) or 127 (UDPv6)      |                    |                  |
++---------------------+--------------------------------------------------------+--------------------+------------------+
```
Comment (JLBuenoLopez-eProsima) [inline docs/fastdds/xml_configuration/common.rst:109]: I suggest adding the `|br|` macro so the table is formatted with line breaks and can be viewed with the webpage margins correctly. Right now, the last two columns are not viewed at first.

File: docs/fastdds/xml_configuration/datawriter.rst [hunk 2]
```diff
@@ -132,12 +132,14 @@ The DataWriter configuration is performed through the XML elements listed in the
     :language: xml
     :start-after: <!-->XML-DATAWRITER<-->
     :end-before: <!--><-->
-    :lines: 2-4, 6-71, 73-74
+    :lines: 2-4, 6-81, 83-84
```
Comment (JLBuenoLopez-eProsima) [inline docs/fastdds/xml_configuration/datawriter.rst:135]: The final lines are not the correct ones (they are also wrong in the current version but I think we can fix that in this PR)

File: docs/fastdds/xml_configuration/domainparticipant.rst [hunk 3]
```diff
@@ -174,12 +174,14 @@ These elements allow the user to define the DomainParticipant configuration.
     :language: xml
     :start-after: <!-->XML-PARTICIPANT<-->
     :end-before: <!--><-->
-    :lines: 2-3, 5-76, 78
+    :lines: 2-3, 5-86, 88
```
Comment (JLBuenoLopez-eProsima) [inline docs/fastdds/xml_configuration/domainparticipant.rst:177]: Again, the end of the XML file is not the correct one (it is not correct in the current documentation version either). I suggest fixing it in this PR

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]
PARTIAL review

File: docs/fastdds/xml_configuration/common.rst (hunk 1 above)
Comment (MiguelCompany) [reply docs/fastdds/xml_configuration/common.rst]: fcb8c80

File: docs/fastdds/xml_configuration/datawriter.rst (hunk 2 above)
Comment (MiguelCompany) [reply docs/fastdds/xml_configuration/datawriter.rst]: 2ab1b34

File: docs/fastdds/xml_configuration/domainparticipant.rst (hunk 3 above)
Comment (MiguelCompany) [reply docs/fastdds/xml_configuration/domainparticipant.rst]: 2ab1b34

**Comment by JLBuenoLopez-eProsima** [comment]
CI is failing: there is an unexpected identation failure and a spelling error (`accross` is spelled with one `c`).

File: docs/fastdds/transport/announced_locators.rst [hunk 4]
```diff
@@ -36,3 +36,25 @@ An external locator is made up of the standard locator fields (kind, address, an
...
+Communication will be performed using the locators of the innermost LAN available.
+
+As an example, consider a network topology where the application is running on a host connected to a LAN of an office,
+which in turn connects to a LAN for all the offices in the same floor, which in turn connects to a LAN for the
+building.
+
+With the default configuration, communication will only occur between hosts on the LAN for the office.
+This is considered the externality level 0, which is reserved for the LANs directly connected to the network interfaces
+of the host where the application is running.
+This is the externality level that will be used on the matching algorithm for the
```
Comment (JLBuenoLopez-eProsima) [inline docs/fastdds/transport/announced_locators.rst:58]: Maybe we can include the reference to the matching algorithm section.

File: docs/fastdds/transport/announced_locators.rst [hunk 5]
```diff
@@ -36,3 +36,25 @@ An external locator is made up of the standard locator fields (kind, address, an
...
+
+The main purpose of the external locators is to enable communication accross different levels of interconnected LANs.
+Communication will be performed using the locators of the innermost LAN available.
+
+As an example, consider a network topology where the application is running on a host connected to a LAN of an office,
+which in turn connects to a LAN for all the offices in the same floor, which in turn connects to a LAN for the
+building.
+
+With the default configuration, communication will only occur between hosts on the LAN for the office.
+This is considered the externality level 0, which is reserved for the LANs directly connected to the network interfaces
```
Comment (JLBuenoLopez-eProsima) [inline docs/fastdds/transport/announced_locators.rst:56]: I think we should clarify that externality 0 is by default and cannot be set by the user

File: docs/fastdds/transport/announced_locators.rst [hunk 6]
```diff
@@ -36,3 +36,25 @@ An external locator is made up of the standard locator fields (kind, address, an
...
+of the host where the application is running.
+This is the externality level that will be used on the matching algorithm for the
+:ref:`default announced locators<default_announced_locators>`.
+
+The floor LAN will be configured as externality level 1.
```
Comment (JLBuenoLopez-eProsima) [inline docs/fastdds/transport/announced_locators.rst:61]: I do not think this should be in its own paragraph. I suggest removing the blank line and keep it in the same, as it is still explaining the same idea.

File: docs/fastdds/transport/announced_locators.rst [hunk 7]
```diff
@@ -58,3 +58,14 @@ The building LAN will be configured as externality level 2.
...
+
+Additional considerations
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Since using external locators increases the number of locators announced, the
+:ref:`allocation limits for locators discovery<remotelocatorsallocationattributes>` would need to be adjusted for
+your application.
+
+Participants running on the same host, but using different addresses on their
+|BuiltinAttributes::metatrafficMulticastLocatorList-qos-api| will discard shared memory transport locators.
```
Comment (JLBuenoLopez-eProsima) [inline docs/fastdds/transport/announced_locators.rst:95]: Is this the correct reference or the one intended is this other one `|BuiltinAttributes::metatraffic_external_unicast_locators-qos-api|`?

File: docs/fastdds/transport/announced_locators.rst [hunk 8]
```diff
@@ -29,7 +29,12 @@ as follows:
 External Locators
 -----------------
 
-The user can configure a set of external locators for each of the lists of unicast locators.
+The user can configure a set of external locators for each of the lists of unicast locators:
+
+* |BuiltinAttributes::metatrafficMulticastLocatorList-qos-api| on :ref:`wireprotocolconfigqos`
```
Comment (JLBuenoLopez-eProsima) [inline docs/fastdds/transport/announced_locators.rst:34]: I do not think this is the correct reference. I am quite sure the one intended here is `|BuiltinAttributes::metatraffic_external_unicast_locators-qos-api|`

File: docs/fastdds/transport/announced_locators.rst [hunk 9]
```diff
@@ -64,6 +64,26 @@ The building LAN will be configured as externality level 2.
 Note that in order for the communication to be successful, routing rules should most probably need to be added to the
 different network routers.
 
+Matching algorithm
+^^^^^^^^^^^^^^^^^^
+
+When a remote entity is discovered, its list of announced locators is processed to select the ones on the innermost
+externality level where the communication can be established.
+
+The highest externality level is checked first.
```
Comment (JLBuenoLopez-eProsima) [inline docs/fastdds/transport/announced_locators.rst:73]: I am not sure if this idea requires its own paragraph...

File: docs/fastdds/transport/announced_locators.rst [hunk 10]
```diff
@@ -64,6 +64,26 @@ The building LAN will be configured as externality level 2.
...
+the same host at that level, and the algorithm proceeds to an inner level.
+If the discovered addresses are not equal to the ones announced by the local entity, processing stops at the current
+level.
+
+When the externality level on which the communication will be established has been decided, the algorithm will:
+* Remove locators that match with addresses on any other externality level.
+* Keep locators that match with the selected externality level.
+* For the locators with an address that does not match with any of the locators announced by the local entity:
+  * Keep them when ``ignore_non_matching_locators`` is ``false`` (default behavior)
+  * Remove them when ``ignore_non_matching_locators`` is ``true``
```
Comment (JLBuenoLopez-eProsima) [inline docs/fastdds/transport/announced_locators.rst:85]: This is the list which indentation is failing

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]

**Review by JLBuenoLopez-eProsima — COMMENTED** [review]
Before approving I would like to check with the author if the references that we have changed are correct

**Comment by MiguelCompany** [comment]
> Before approving I would like to check with the author if the references that we have changed are correct

I confirm they are correct

**Pull request merged**